lib/support/wercker/utils/wercker_cache_handler.sh

```diff
@@ -0,0 +1,223 @@
+#!/bin/bash
+# HANDLES CACHE FOR SYSTEM LEVEL DEPENDENCIES
+set -e
+
+# TODO: invalidate cache in case of errors
+# TODO: make somehow independent from the operating system, installing the tools we use
+# TODO: improve command output colors using the Wercker library to do so 
+# TODO: make it more generic, especially the methods handling store/restore of the cache
+#				because now they somehow are designed specifically for the use we to in ASQ,
+#	      especially for what concern the excluded folders
+
+#$1 = cache_folder_name
+#$2 = file based on which we decide if invalidate the cache, and that has to be run
+#$3 = space separated list of folders we want to cache
+cache() {
+	work_with_cache $1 $2 "$3";
+}
+
+#$1 = cache_folder_name
+#$2 = file based on which we decide if invalidate the cache, and that has to be run
+#$3 = space separated list of folders we want to cache
+work_with_cache() {
+	# if it is the first time we are using the cache
+	if [ ! -d "$WERCKER_CACHE_DIR/$1" ] || [ ! -f "$WERCKER_CACHE_DIR/$1/$2" ]; then
+		echo "First time initialization of the cache for "$1
+		setup_cache $1 $2;
+		execute_file $2;
+		handle_cache "store" $1 "$3";
+		adds_file_invalidating_the_cache $1 $2;
+	else
+  	cache_status=$(check_cache_validity $1 $2)
+
+  	# Debug echoes
+  	# NOTE: The comment are here, because they do not get printed on
+  	#       Wercker if inside the subshell call
+  	md5_current=($(md5sum "$2"))
+		md5_cached=($(md5sum "$WERCKER_CACHE_DIR/$1/$2"))
+		echo "Cached MD5 for "$2" is "$md5_cached
+		echo "Current MD5 for "$2" is "$md5_current
+
+  	# if the cache it is not valid
+  	if [ "$cache_status" = "invalid" ]; then
+  		echo "Cache is invalid for "$1
+  		clear_cache $1;
+  		setup_cache $1 $2;
+			execute_file $2;
+			adds_file_invalidating_the_cache $1 $2;
+			handle_cache "store" $1 "$3";
+		else
+			echo "Reusing cache for "$1
+			handle_cache "restore" $1 "$3";
+		fi
+
+	fi
+
+}
+
+#$1 = cache_folder_name
+setup_cache() {
+
+	echo "Setting up the cache for "$1
+
+	if [ ! -d "$WERCKER_CACHE_DIR/$1" ]; then
+	  echo 'Creating '$WERCKER_CACHE_DIR'/'$1
+  	mkdir -p "$WERCKER_CACHE_DIR/$1"
+  else
+  	echo 'Already existing '$WERCKER_CACHE_DIR'/'$1
+	fi
+
+	echo "Done Setting up cache for "$1
+}
+
+#$1 = action: store or restore
+#$2 = cache_folder_name
+#$3 = space separated list of folders we want to cache
+handle_cache() {
+	folders_array=($3)
+	number_of_folders=${#folders_array[*]}
+
+	if [ "$number_of_folders" -eq "1" ]; then
+		handle_single_folder_cache $1 $2 "$3";
+	else
+		handle_multiple_folder_cache $1 $2 "$3";
+	fi
+}
+
+# TODO: refactor to remove duplicated code across methods handling the cache
+#$1 = action: store or restore
+#$2 = cache_folder_name
+#$3 = space separated list of folders we want to cache
+handle_multiple_folder_cache() {
+	for directory in $3; do
+
+		directory_name_with_underscore=${directory//\//\_}
+
+		case $1 in
+	    "store")
+	      SRC=$WERCKER_SOURCE_DIR'/'${directory}
```

Comment:
Not an shell script expert, but do you guard against spaces. Check the "Be prepared for spaces in filenames"  section here: http://www.davidpashley.com/articles/writing-robust-shell-scripts/